src/dtx/dtx_internal.h

```diff
@@ -192,6 +192,14 @@ extern uint32_t dtx_batched_ult_max;
  */
 #define DTX_COLL_TREE_WIDTH		8
 
+/*
+ * If a large transaction has sub-requests to dispatch to a lot of DTX participants,
+ * then we may have to split the dispatch process to multiple steps; otherwise, the
+ * dispatch process may trigger too many in-flight or in-queued RPCs that will hold
+ * too much resource as to server maybe out of memory.
+ */
+#define DTX_RPC_STEP_LENGTH	DTX_THRESHOLD_COUNT
+
```

Comment:
Is it feasible to have arbitrary limits like this in production?
At one end, what if the limit is too high and still floods the server?
At the other end, what if the limit is too low and we don't utilize the available bandwidth / performance?
Tuning things like this for each system seems difficult and error-prone. Should/are we looking at ways to handle this dynamically? And not just THIS limit, but many other arbitrary limits and thresholds that exist

Reply:
Here, the things to be resolved is that on a large system, such as Aurora, if a lot of engines send out massive RPCs to the same engine in a short time, it may cause the recv_buffer on the destination to be exhausted. So controlling the RPCs load peak will be helpful to reduce such possibility. On the other hand, if one engine sends out too many RPCs to the others concurrently, these in-flight RPCs will occupy a lot of DRAM as to DRAM pressure on the engine. That will affect others (on the same engine) to dynamically allocate DRAM. So the affect of RPC load peak is not only globally but also locally.

There are many factors to be considered. It is not expected that all the DAOS-15914 issues can be resolved if only applying this patch without other (network configuration) changes. On the other hand, this patch only controls DTX batched commit RPCs. That is server side background task, and will not direct affect client side IO bandwidth.

Honestly, this "limit" is some kinds of experience value instead of a strict calculated one. It will be great if we can automatically adjust this "limit" based on the system detailed configuration in the future (although that is not easy). But it does not conflict with current experience value based first-step.

Reply:
My concern is that every time we scale higher on Aurora, we hit new limits. We're seeing issues like this with less than 10% of the servers, and less than 5% of the computes. So this is still "small" scale.
I'm not picking on THIS PR specifically. It's more a general concern that repeatedly bumping limits and configuring envs might not be practical. Even if we arrive at a "good" config, it's very likely that new workloads in production will blow up the system again.

While we're debugging these issues now, and learning from them, this is okay. But in production we can't ask Admins to restart and try different patches and configs every time. So if we work towards more general and dynamic solutions, we can safeguard against future unknown workloads.

Reply:
This patch is not a debug patch to increase or decrease the "threshold", instead, such "threshold" has already been there for sometime. Originally, we used such "threshold" to restrict IO forwarded RPCs workload; now, we extend the restriction to cover DTX batched commit RPCs.

I can understand your concern. If running DAOS on more large scaled cluster, whether the "threshold" will be changed or not. Honestly, I do not know. But I tend to say "no". If we make the "threshold" to be configurable, it will be difficult for the user to determine how large the "threshold" will be a reasonable value. So be as the first step, keeping it as a fixed value will be the most simple solution. If we can automatically adjust such "threshold" based on the system (instead of only current DAOS engine) resource pressure in the future, then we can replace such fixed "threshold" without any compatibility issues. Such dynamical adjustment is more related with RPC flow control among DAOS engines. That is not small work.